src/mopup/__init__.py

```diff
@@ -202,11 +317,384 @@ def do_download(download_url: DecodedURL) -> str:
+            partialdir.rmdir()
+            return str(finalname)
+
+
+def _ensure_sudo_if_needed(dry_run: bool) -> None:
+    """Ensure we have sudo privileges if needed, or relaunch with sudo."""
+    is_root = geteuid() == 0
+
+    # If not root and not dry_run, relaunch with sudo
+    if not is_root and not dry_run:
+        # Check if sudo is already active
+        sudo_check = run(  # noqa: S603
+            ["/usr/bin/sudo", "-nv"],
+            capture_output=True,
+        )
+
+        # If sudo is not active, enable it
+        if sudo_check.returncode != 0:
+            print("Enter your administrative password to uninstall:")
+            result = run(["/usr/bin/sudo", "-v"])  # noqa: S603
+            if result.returncode != 0:
+                print("Failed to obtain administrative privileges.")
+                sys.exit(1)
+
+        # Relaunch ourselves with sudo
+        cmd = ["/usr/bin/sudo", executable, "-BI"] + argv
+        result = run(cmd, cwd="/")  # noqa: S603
+
+        if sudo_check.returncode != 0:
+            # Kill sudo session after completion if we enabled it
+            run(["/usr/bin/sudo", "-k"])  # noqa: S603
+            sys.exit(result.returncode)
+
+
+def _find_python_packages(major: str, minor: str) -> list[str]:
+    """Find all Python packages for a specific version."""
+    result = run(  # noqa: S603
+        ["/usr/sbin/pkgutil", "--pkgs"],
+        stdout=PIPE,
+        text=True,
+    )
+
+    packages = [
+        pkg
+        for pkg in result.stdout.strip().split("\n")
+        if pkg.startswith("org.python.Python.") and pkg.endswith(f"-{major}.{minor}")
+    ]
+
+    return packages
+
+
+def _collect_package_files(
+    packages: list[str],
+) -> tuple[set[Path], set[Path], set[Path]]:
+    """Collect all files and directories from packages."""
+    all_files: set[Path] = set()
+    all_dirs: set[Path] = set()
+    package_root_dirs: set[Path] = set()
+
+    # Extract version from package names (e.g. "org.python.Python.PythonFramework-3.13")
+    version = None
+    for pkg in packages:
+        if "-" in pkg:
+            version = pkg.split("-")[-1]
+            break
+
+    for pkg in packages:
+        # Export package info as plist and convert to JSON to preserve \r characters
+        # (plistlib's XML parsing normalizes \r to \n, but JSON preserves them)
+        plist_result = run(  # noqa: S603
+            ["/usr/sbin/pkgutil", "--export-plist", pkg],
+            capture_output=True,
+        )
+
+        if plist_result.returncode != 0:
+            print(f"Warning: Failed to get pkgutil plist for {pkg}", end="")
+            print(f":{err.decode()}" if (err := plist_result.stderr) else "")
+            continue
+
+        # Convert plist XML to JSON using plutil
+        json_result = run(  # noqa: S603
+            ["/usr/bin/plutil", "-convert", "json", "-o", "-", "-"],
+            input=plist_result.stdout,
+            capture_output=True,
+        )
```

Comment:
This is unnecessary; you can load it directly with [`plistlib.load`](https://docs.python.org/3.13/library/plistlib.html#plistlib.loads)

Reply:
(plistlib uses expat internally)